### PR TITLE
[asl][semantics reference] Addressing feedback

### PR DESCRIPTION
**Pull request by Roman-Manevich**
- [x] Addressing all feedback comments.
- [x] Folding Reading Guide into the Introduction chapters of both ASL Typing Reference and ASL Semantics Reference.
- [x] Ensuring mutual exclusion of the various assignment rules.
- [x] Removed all mention of the LRM (conditionally compiled away).

### CODE REVIEW

File: asllib/doc/ASLSemanticsReference.tex [hunk 1]
```diff
@@ -357,7 +362,8 @@ \section{When Do ASL Specifications Have Meaning \label{sec:MeaningfulASLSpecifi
 
 Although ASL does not require specifications to terminate, the semantics defined in this
 document assign meaning only to terminating specifications.
-A future version of this document, will assign meaning to potentially non-terminating specifications.
+A future version of this document, will assign meaning to potentially non-terminating specifications,
+potentially adopting the approach of Ancona et al.~\cite{AnconaDRZ20}.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:366]: please remove the "potentially" phrase

File: asllib/doc/ASLSemanticsReference.tex [hunk 2]
```diff
@@ -967,95 +971,87 @@ \chapter{Evaluation of Expressions \label{chap:eval_expr}}
...
 \end{emptyformal}
 
-  \subsection{Comments}
-  When there exists a global variable $\vx$, the type system
-  forbids having $\vx$ as a local variable.
-  This is enforced by TypingRule.LDVar and TypingRule.DeclareOneFunc.
+\subsection{Comments}
+When there exists a global variable $\vx$, the type system
+forbids having $\vx$ as a local variable.
+This is enforced by TypingRule.LDVar and TypingRule.DeclareOneFunc.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:1033]: please add a reference to the Typing Reference, with pointers to the specific chapters

File: asllib/doc/ASLSemanticsReference.tex [hunk 3]
```diff
@@ -1101,28 +1095,26 @@ \section{SemanticsRule.BinopAnd \label{sec:SemanticsRule.BinopAnd}}
 \end{mathpar}
 \end{emptyformal}
 
-  \subsection{Comments}
-  The evaluation via the rule above ensures that $\veone$ is evaluated first and only if
-  it evalutes to \True, is $\vetwo$ evaluated.
+\subsection{Comments}
+The evaluation via the rule above ensures that $\veone$ is evaluated first and only if
+it evaluates to \True, is $\vetwo$ evaluated.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:1100]: typo: is vetwo evaluated?

File: asllib/doc/ASLSemanticsReference.tex [hunk 4]
```diff
@@ -1237,24 +1228,20 @@ \subsection{Comments}
...
-  \end{itemize}
-
-  \subsection{Example}
-    In the specification:
-    \VerbatimInput{../tests/ASLSemanticsReference.t/SemanticsRule.EUnopAssert.asl}
-    the expression \texttt{NOT '1010'} evaluates to the value \texttt{'0101'}.
+\subsection{Prose}
+All of the following apply:
+\begin{itemize}
+\item $\ve$ denotes a Unary Operator $\op$ over an expression, $\EUnop(\op, \veone)$;
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:1234]: Unary Operator -> unary operator

File: asllib/doc/ASLSemanticsReference.tex [hunk 5]
```diff
@@ -1237,24 +1228,20 @@ \subsection{Comments}
...
+\item the evaluation of the expression $\veone$ in $\env$ yields \\ $\Normal((\vvone, \vg), \newenv)$\ProseOrAbnormal;
+\item applying the Unary Operator $\op$ to $\vvone$ is $\vv$.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:1236]: Unary Operator -> unary operator

File: asllib/doc/ASLSemanticsReference.tex [hunk 6]
```diff
@@ -1371,30 +1350,28 @@ \section{SemanticsRule.ESlice \label{sec:SemanticsRule.ESlice}}
...
-    \item one of the following apply:
+    \item all of the following apply (\textsc{single\_returned\_value}):
     \begin{itemize}
-      \item all of the following apply (\textsc{single\_returned\_value}):
-      \begin{itemize}
-        \item $\vms$ consists of a single returned value $(\vv,\vg)$,
-        which go into the output configuration $\Normal((\vv, \vg), \newenv)$.
-      \end{itemize}
+      \item $\vms$ consists of a single returned value $(\vv,\vg)$,
+      which go into the output configuration $\Normal((\vv, \vg), \newenv)$.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:1363]: which goes?

File: asllib/doc/ASLSemanticsReference.tex [hunk 7]
```diff
@@ -1484,32 +1458,28 @@ \section{SemanticsRule.EGetArray \label{sec:SemanticsRule.EGetArray}}
 \isempty{\subsection{Comments}}
 
 \section{SemanticsRule.ERecord \label{sec:SemanticsRule.ERecord}}
+\subsection{Prose}
+All of the following apply:
+\begin{itemize}
+\item $\ve$ denotes a record creation expression, $\ERecord(\names, \efields)$;
+\item the names of of the fields are $\id_{1..k}$;
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:1465]: typo: of of

File: asllib/doc/ASLSemanticsReference.tex [hunk 8]
```diff
@@ -2127,22 +2069,20 @@ \section{SemanticsRule.LESetArray \label{sec:SemanticsRule.LESetArray}}
 \end{emptyformal}
 
 \subsection{Comments}
-  This is related to \identr{WHRS}.
+\lrmcomment{This is related to \identr{WHRS}.}
 
-  We note that the index is guaranteed by the type-checker to be within the array bounds
-  via TypingRule.LESetArray.
+We note that the index is guaranteed by the type-checker to be within the array bounds
+via TypingRule.LESetArray.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:2075]: please add pointer to the typing ref, with the specific chapter

File: asllib/doc/ASLSemanticsReference.tex [hunk 9]
```diff
@@ -2178,16 +2118,14 @@ \section{SemanticsRule.LESetField \label{sec:SemanticsRule.LESetField}}
 \end{emptyformal}
 
 \subsection{Comments}
-  This is related to \identr{WHRS}.
+\lrmcomment{This is related to \identr{WHRS}.}
 
-  We note that the type-check guarantees that $\fieldname$ exists in the record given by $\record$
-  via TypingRule.LESetStructuredField.
+We note that the type-check guarantees that $\fieldname$ exists in the record given by $\record$
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:2123]: type-checker

File: asllib/doc/ASLSemanticsReference.tex [hunk 10]
```diff
@@ -2196,18 +2134,19 @@ \section{SemanticsRule.LEDestructuring \label{sec:SemanticsRule.LEDestructuring}
     \item $\nmonads$ is the list of pairs consisting of $\vv_i$ and $\vg$ for $i=1..k$;
     \item evaluating the multi-assignment between $\vlelist$ and the list $\nmonads$
     in $\env$ achieves the effects of assigning each value to the respective
-    subexpressions, resulting in the output configuration $C$.
+    sub-expressions, resulting in the output configuration $C$.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:2137]: please drop the hyphen

File: asllib/doc/ASLSemanticsReference.tex [hunk 11]
```diff
@@ -3204,63 +3113,67 @@ \chapter{Evaluation of Statements \label{chap:eval_stmt}}
...
-  \begin{itemize}
-  \item $\vs$ is a \texttt{pass} statement, $\SPass$;
-  \item $\newg$ is the empty graph;
-  \item $\newenv$ is $\env$.
-  \end{itemize}
+\subsection{Prose}
+All of the following apply:
+\begin{itemize}
+\item $\vs$ is a \texttt{pass} statement, $\SPass$;
+\item define $\newg$ as the empty graph;
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:3120]: please rephrase as it was before "newg is the empty graph", or expand

File: asllib/doc/ASLSemanticsReference.tex [hunk 12]
```diff
@@ -3204,63 +3113,67 @@ \chapter{Evaluation of Statements \label{chap:eval_stmt}}
...
+\item define $\newenv$ as $\env$.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:3121]: same here

File: asllib/doc/ASLSemanticsReference.tex [hunk 13]
```diff
@@ -3204,63 +3113,67 @@ \chapter{Evaluation of Statements \label{chap:eval_stmt}}
...
 \end{emptyformal}
 
 \isempty{\subsection{Comments}}
 
 \section{SemanticsRule.SAssign \label{sec:SemanticsRule.SAssign}}
-  \subsection{Prose}
-  Evaluation of the statement $\vs$ in an environment $\env$ is \\
-  $\Continuing(\newg, \newenv)$ and all of the following apply:
-  \begin{itemize}
+This rule covers all assignment statements, except ones where the right-hand side expression
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:3143]: except the ones?

File: asllib/doc/ASLSemanticsReference.tex [hunk 14]
```diff
@@ -3270,30 +3183,31 @@ \section{SemanticsRule.SAssign \label{sec:SemanticsRule.SAssign}}
...
-  \end{itemize}
+\end{itemize}
 
-    \subsection{Example}
+\subsection{Example}
 
 \VerbatimInput{\testdir/SemanticsRule.SAssignCall.asl}
-  given that the function call \texttt{f(1)} returns the pair of values \texttt{(1,2)}, statement \texttt{(a,b) = f(1)} assigns the value \texttt{1} to the mutable variable \texttt{a} and the value \texttt{2} to the mutable variable~\texttt{b}.
+given that the function call \texttt{f(1)} returns the pair of values \texttt{(1,2)},
+statement \texttt{(a,b) = f(1)} assigns the value \texttt{1} to the mutable variable \texttt{a}
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:3204]: the statement

File: asllib/doc/ASLSemanticsReference.tex [hunk 15]
```diff
@@ -3459,19 +3369,13 @@ \section{SemanticsRule.SReturnSome \label{sec:SemanticsRule.SReturnSome}}
...
   \begin{itemize}
     \item $\vs$ is a \emph{sequencing statement} \texttt{s1; s2}, that is, $\SSeq(\vsone, \vstwo)$;
     \item evaluating $\vsone$ in $\env$ is either $\Continuing(\vgone, \envone)$ in which case
     the evaluation continues,
-    or a returning configuration ($\Returning((\vvs, \newg), \newenv)$) or an abnormal configuration,
-    which short-circuit the entire evaluation;
-    \item evaluating $\vstwo$ in $\envone$ is either a non-abnormal configuration $C$
-    \ProseOrAbnormal;
+    or a returning configuration ($\Returning((\vvs, \newg), \newenv)$)\ProseOrAbnormal;
+    \item evaluating $\vstwo$ in $\envone$ yields a non-abnormal configuration $C$\ProseOrAbnormal;
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:3378]: non-abnormal = normal?

File: asllib/doc/ASLSemanticsReference.tex [hunk 16]
```diff
@@ -3543,13 +3445,12 @@ \section{SemanticsRule.SCond \label{sec:SemanticsRule.SCond}}
...
     \item $\vs$ is a condition statement, $\SCond(\ve, \vsone, \vstwo)$;
-    \item evaluating $\ve$ in $\env$ is either $\Normal((\vv, \vgone)$ \ProseOrAbnormal;
+    \item evaluating $\ve$ in $\env$ is $\Normal((\vv, \vgone)$\ProseOrAbnormal;
     \item $\vv$ is a native Boolean for $\vb$;
     \item the statement $\vsp$ is $\vsone$ is $\vb$ is $\True$ and $\vstwo$ otherwise
     (so that $\vsone$ will be evaluated if the condition evaluates to $\True$ and otherwise
     $\vstwo$ will be evaluated);
-    \item evaluating $\vsp$ in $\envone$ as per \chapref{eval_block} is either
-    a configuration $C$, which is not abnormal \ProseOrAbnormal;
+    \item evaluating $\vsp$ in $\envone$ as per \chapref{eval_block} is a non-abnormal configuration $C$\ProseOrAbnormal;
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:3453]: non-abnormal = normal?

File: asllib/doc/ASLSemanticsReference.tex [hunk 17]
```diff
@@ -3643,7 +3544,7 @@ \section{SemanticsRule.SCase \label{sec:SemanticsRule.SCase}}
   }
 \end{mathpar}
 
-We now define the semantics of a case statement in terms of the desugared statement:
+We now define the semantics of a case statement in terms of the de-sugared statement:
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:3547]: drop the hyphen

File: asllib/doc/ASLSemanticsReference.tex [hunk 18]
```diff
@@ -3892,29 +3783,26 @@ \section{SemanticsRule.SThrowNone \label{sec:SemanticsRule.SThrowNone}}
...
+  \item $\newg$ is the ordered composition of $\vgone$ and $\vgtwo$ with the $\asldata$ edge;
+  \item $\vex$ consists of the exception value $\vv$, the name of the variable holding it ---
+        $\name$, and the type annotation for the exception --- $\vt$;
+  \item the result of the entire evaluation is $\Throwing((\vex, \newg), \env)$.
+\end{itemize}
 
+\subsection{Example}
+The specification:
+\VerbatimInput{../tests/ASLSemanticsReference.t/SemanticsRule.SThrowSomeTyped.asl}
+terminates successfully. That is, not dynamic error occurs.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:3803]: not -> no

File: asllib/doc/ASLSemanticsReference.tex [hunk 19]
```diff
@@ -3937,13 +3825,11 @@ \subsection{Formally}
...
     \subsection{Prose}
-    Evaluation of the statement $\vs$ is either a non-abnormal configuration $C$
-    or an abnormal configuration,
-    and all of the following apply:
+    All of the following apply:
     \begin{itemize}
     \item $\vs$ is a \texttt{try} statement, $\STry(\vs, \catchers, \otherwiseopt)$;
     \item evaluating $\vsone$ in $\env$ as per \chapref{eval_block}
-    is either a non-abnormal configuration $\sm$ \ProseOrAbnormal;
+    is a non-abnormal configuration $\sm$\ProseOrAbnormal;
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:3832]: non-abnormal = normal?

File: asllib/doc/ASLSemanticsReference.tex [hunk 20]
```diff
@@ -4285,19 +4158,14 @@ \subsubsection{Stepping the Index Variable}
...
-  is either \\ $\Continuing(\vgone, \envone)$ or an early return configuration or an abnormal configuration,
-  which short-circuits the entire evaluation;
+  is \\ $\Continuing(\vgone, \envone)$\ProseTerminateAs{\ReturningConfig, \ThrowingConfig, \ErrorConfig};
   \item stepping the index $\vindexname$ with $\vstart$ and the direction $\dir$ in $\envone$,
   that is, $\evalforstep(\envone, \vindexname, \vstart, \dir)$ gives $((\vstep, \envtwo), \vgtwo)$;
   \item evaluating the \texttt{for} loop with $(\vindexname, \vstep, \dir, \vend, \vbody)$
-  in $\envtwo$ wither results in a continuing configuration $\Continuing(\vgthree, \newenv)$
-  or an early return configuration
-  or an abnormal configuration, which short-circuits the entire evaluation;
+  in $\envtwo$ wither results in a continuing configuration $\Continuing(\vgthree, \newenv)$\ProseTerminateAs{\ReturningConfig, \ThrowingConfig, \ErrorConfig};
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:4168]: wither?

File: asllib/doc/ASLSemanticsReference.tex [hunk 21]
```diff
@@ -4492,42 +4359,41 @@ \chapter{Evaluation of Catchers \label{chap:eval_catchers}}
...
   \item $\envthrow$ consists of the static environment $\tenv$ and dynamic environment \\ $\denvthrow$;
   \item $\envone$ is defined by taking the static environment $\tenv$, the global component of the dynamic
-  environment from $\denvthrow$ and the local component of the dynamic environment from $\denv$;
+        environment from $\denvthrow$ and the local component of the dynamic environment from $\denv$;
   \item finding the first catcher with the static environment $\tenv$, the exception type $\vvty$,
   and the list of catchers $\catchers$ gives a catcher that does not declare a name ($\None$) and gives a statement $\vs$;
-  \item evaluating $\vs$ in $\envone$ as a block (\chapref{eval_block}) is either a non-error
-  configuration $C$ \ProseOrError;
+  \item evaluating $\vs$ in $\envone$ as a block (\chapref{eval_block}) is a non-error
+        configuration $C$\ProseOrError;
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:4373]: is a non-error -> is not an error

File: asllib/doc/ASLSemanticsReference.tex [hunk 22]
```diff
@@ -4552,8 +4418,8 @@ \section{SemanticsRule.CatchNamed \label{sec:SemanticsRule.CatchNamed}}
     and the list of catchers $\catchers$ gives a catcher that declares the name $\name$ and gives a statement $\vs$;
     \item $\vgone$ is the execution graph resulting from reading $\vv$ into the identifier $\eid$;
     \item declaring a local identifier $\name$ with $(\veone, \vgone)$ in $\envone$ gives $(\envtwo, \vgtwo)$;
-    \item evaluating $\vs$ in $\envtwo$ asa block (\chapref{eval_block}) is either a non-error
-    configuration $C$ \ProseOrError;
+    \item evaluating $\vs$ in $\envtwo$ asa block (\chapref{eval_block}) is a non-error
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:4421]: asa -> as a
is a non-error -> is not an error

File: asllib/doc/ASLSemanticsReference.tex [hunk 23]
```diff
@@ -4612,7 +4478,7 @@ \section{SemanticsRule.CatchOtherwise \label{sec:SemanticsRule.CatchOtherwise}}
     and the list of catchers $\catchers$ gives a catcher that declares the name $\name$ and gives $\None$
     (that is, neither of the \texttt{catch} clauses matches the raised exception);
     \item evaluating the \texttt{otherwise} statement $\vs$ in $\envtwo$ asa block (\chapref{eval_block})
-    is either a non-error configuration $C$ \ProseOrError;
+    is a non-error configuration $C$\ProseOrError;
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:4481]: is a non-error -> is not an error

File: asllib/doc/ASLSemanticsReference.tex [hunk 24]
```diff
@@ -5113,21 +4979,21 @@ \section{SemanticsRule.FPrimitive \label{sec:SemanticsRule.FPrimitive}}
...
-  \subsection{Formally}
+\subsection{Formally}
   The following rule utilizes the transition relation
   \hypertarget{def-evalprimitive}{}
   \[
     \evalprimitive{\overname{\Identifiers}{\name} \aslsep \overname{(\vals\times\XGraphs)^*}{\actualargs}} \bigtimes
     \Normal(\overname{(\vals\times\XGraphs)^*}{\vms} \aslsep \overname{\XGraphs}{\vgone}) \cup \overname{\TError}{\ErrorConfig} \enspace,
   \]
-  which parameterizes the ASL semantics and allows evaluating primitive subprograms.
+  which parametrizes the ASL semantics and allows evaluating primitive subprograms.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:4989]: 3rd person present: parameterizes

File: asllib/doc/ASLSemanticsReference.tex [hunk 25]
```diff
@@ -6401,7 +6265,7 @@ \subsection{Prose}
   (\overname{\Bool}{\vb} \times \overname{\XGraphs}{\vg}) \cup \overname{\TError}{\ErrorConfig}
 \]
 checks whether the value $\vv$ can be stored in a variable of type $\vt$ in the environment $\env$,
-resulting in a Boolean value $\vb$ and execution graph $\vg$, \ProseOrError.
+resulting in a Boolean value $\vb$ and execution graph $\vg$ or a dynamic error.
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:6268]: why drop the macro here?

File: asllib/doc/ASLSemanticsReference.tex [hunk 26]
```diff
@@ -6701,6 +6558,7 @@ \section{SemanticsRule.BinopOperator \label{sec:SemanticsRule.BinopOperator}}
   \binopliterals(\overname{\binop}{\op} \aslsep \overname{\literals}{\vvone} \aslsep \overname{\literals}{\vvtwo}) \aslto
   \overname{\literals}{\vr} \cup \TTypeError
 \]
+which is defined in the ASL Typing Reference~\cite{ASLTypingReference} (see TypingRule.BinopLiterals).
```
Comment (jalglave) [inline asllib/doc/ASLSemanticsReference.tex:6561]: please also point to the right chapter

**Review by jalglave — COMMENTED** [review]
thanks very much Roman, here are a few comments

File: asllib/doc/ASLSemanticsReference.tex (hunk 3 above)
Comment (Roman-Manevich) [reply asllib/doc/ASLSemanticsReference.tex:1100]: fixed

File: asllib/doc/ASLSemanticsReference.tex (hunk 15 above)
Comment (Roman-Manevich) [reply asllib/doc/ASLSemanticsReference.tex:3378]: can also be continuing. I added explanation.

File: asllib/doc/ASLSemanticsReference.tex (hunk 16 above)
Comment (Roman-Manevich) [reply asllib/doc/ASLSemanticsReference.tex:3453]: same fix

File: asllib/doc/ASLSemanticsReference.tex (hunk 19 above)
Comment (Roman-Manevich) [reply asllib/doc/ASLSemanticsReference.tex:3832]: same fix as bove

File: asllib/doc/ASLSemanticsReference.tex (hunk 20 above)
Comment (Roman-Manevich) [reply asllib/doc/ASLSemanticsReference.tex:4168]: fixed

File: asllib/doc/ASLSemanticsReference.tex (hunk 25 above)
Comment (Roman-Manevich) [reply asllib/doc/ASLSemanticsReference.tex:6268]: this macro should only appear in premises.

**Pull request merged**